tap_s3_csv/dialect.py

```diff
@@ -0,0 +1,180 @@
+import re
+import random
+import csv
+import chardet
+import clevercsv
+
+from tap_s3_csv import s3
+
+def detect_tables_dialect(config):
+    # there is only one table in the array
+    for table in config['tables']:
+        # will return all matching files in s3 with given prefix and table name in config
+        s3_files = s3.get_input_files_for_table(config, table)
+
+        for s3_file in s3_files:
+            detect_dialect(config, s3_file, table)
+
+
+def detect_dialect(config, s3_file, table):
+    config_delimiter = table.get('delimiter', '')
+    config_quotechar = table.get('quotechar', '')
+    config_encoding = table.get('encoding', '')
+
+    detect_delimiter = config_delimiter == ''
+    detect_quotechar = config_quotechar == ''
+    detect_encoding = config_encoding == ''
+
+    if not detect_delimiter and not detect_delimiter and not detect_quotechar:
+        return
+
+    # clevercsv is good but slow - we cap it at 2000 rows, which is 1s of runtime on my machine
+    MAX_DIALECT_LINES = 2000
+    MAX_ENCODING_LINES = 10000
+    MAX_LINES = MAX_ENCODING_LINES if detect_encoding else MAX_DIALECT_LINES
+
+    # max bytes we want to cache in memory
+    MAX_LINES_BYTES = 25 * 1024 ** 2
+
+    # max bytes for each line read
+    MAX_LINE_BYTES = 1024 ** 2
+
+    # chardet is slow and rarely detects early. We limit the number of lines it is fed to keep performance acceptable.
+    # The question is how many lines and how do we pick the most interesting lines?
+    #
+    # The answer below is the DETECT_CHARDET_LINE. The intuition is a line with all ASCII characters is unlikely to
+    # change chardet's mind on anything. We took the HIGH_BYTE_DETECTOR and ESC_DETECTOR regexs from chardet and
+    # combined them to filter for lines that are worth feeding into chardet. We can search through a large number of
+    # lines for interesting ones quickly since chardet isn't involved. This rationale seems better than first x lines.
+    #
+    # We always take the first line since it may have BOM information and it's often the header. We think it's good
+    # to take some of the beginning lines since if it fails to detect and the encoding is obvious to a person by
+    # looking at the first few lines, it's especially offensive. People might be more understanding of a failed
+    # detection if the key line is buried deep in the file.
+    DETECT_CHARDET_LINE = re.compile(b'([\x80-\xFF]|(\033|~{))')
+    MAX_CHARDET_LINES = 100
+    FIRST_CHARDET_LINES = MAX_CHARDET_LINES / 10
+    interesting = []
+    interesting_map = {}
+
+    lines = []
+    lines_read = 0
+
+    file_key = s3_file.get('key')
+    file_handle = s3.get_file_handle(config, file_key)
+    file_iter = file_handle.iter_lines()
```

Comment:
does calling this download the entire file? tap-s3-csv seems to read files differently with raw_stream and such.